pinot-core/src/main/java/org/apache/pinot/core/query/reduce/GroupByDataTableReducer.java

```diff
@@ -171,106 +160,81 @@ public void reduceAndSetResults(String tableName, DataSchema dataSchema,
+      String[] reorderedColumnNames = new String[numSelectExpressions];
+      DataSchema.ColumnDataType[] reorderedColumnDataTypes = new DataSchema.ColumnDataType[numSelectExpressions];
+      resultTableSchema = new DataSchema(reorderedColumnNames, reorderedColumnDataTypes);
+      for (int i = 0; i < numSelectExpressions; i++) {
+        reorderedColumnNames[i] = columnNames[selectExpressionIndexMap[i]];
+        reorderedColumnDataTypes[i] = columnDataTypes[selectExpressionIndexMap[i]];
       }
-      if (_sqlSelectionList != null) {
-        Object[] finalValues = new Object[_sqlSelectionList.size()];
-        for (int i = 0; i < finalSchemaMapIdx.length; i++) {
-          finalValues[i] = values[finalSchemaMapIdx[i]];
+      while (rows.size() < limit && sortedIterator.hasNext()) {
+        Record nextRecord = sortedIterator.next();
+        Object[] values = nextRecord.getValues();
+        for (int i = 0; i < _numAggregationFunctions; i++) {
+          int valueIndex = i + _numGroupByExpressions;
+          values[valueIndex] = AggregationFunctionUtils
+              .getSerializableValue(_aggregationFunctions[i].extractFinalResult(values[valueIndex]));
         }
-        rows.add(finalValues);
-      } else {
-        rows.add(values);
-      }
-      numRows++;
-    }
 
-    DataSchema finalDataSchema = getSQLResultTableSchema(dataSchema);
-    if (_sqlSelectionList != null) {
-      int columnSize = _sqlSelectionList.size();
-      String[] columns = new String[columnSize];
-      DataSchema.ColumnDataType[] columnDataTypes = new DataSchema.ColumnDataType[columnSize];
-      for (int i = 0; i < columnSize; i++) {
-        columns[i] = finalDataSchema.getColumnName(finalSchemaMapIdx[i]);
-        columnDataTypes[i] = finalDataSchema.getColumnDataType(finalSchemaMapIdx[i]);
+        Object[] reorderedValues = new Object[numSelectExpressions];
+        for (int i = 0; i < numSelectExpressions; i++) {
+          reorderedValues[i] = values[selectExpressionIndexMap[i]];
+        }
+        rows.add(reorderedValues);
       }
-      finalDataSchema = new DataSchema(columns, columnDataTypes);
-    }
-    brokerResponseNative.setResultTable(new ResultTable(finalDataSchema, rows));
-  }
-
-  /**
-   * Generate index mapping based on selection expression to DataTable schema, which is groupBy columns,
-   * then aggregation functions.
-   *
-   * @return a mapping from final schema idx to corresponding idx in data table schema.
-   */
-  private int[] getFinalSchemaMapIdx() {
-    int[] finalSchemaMapIdx = new int[_sqlSelectionList.size()];
-    int nextAggregationIdx = _numGroupBy;
-    for (int i = 0; i < _sqlSelectionList.size(); i++) {
-      finalSchemaMapIdx[i] = getExpressionMapIdx(_sqlSelectionList.get(i), nextAggregationIdx);
-      if (finalSchemaMapIdx[i] == nextAggregationIdx) {
-        nextAggregationIdx++;
+    } else {
+      while (rows.size() < limit && sortedIterator.hasNext()) {
```

Comment:
It will be good to add a comment here stating that this is for the code path for PQL query compiled by PQLCompiler but having both the options (groupByModeSql and responseFormatSqL to true).

Reply:
Added